<a href="https://colab.research.google.com/github/Charles-Scott-Green/Argument-mining/blob/master/3_Baseline_Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
!python -m spacy download en_core_web_lg --quiet

     |████████████████████████████████| 826.9MB 63.3MB/s 
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_lg')


In [0]:
!pip install vaderSentiment

     |████████████████████████████████| 133kB 2.7MB/s 


In [0]:
!pip install spacy-wordnet

     |████████████████████████████████| 655kB 2.7MB/s 
     |████████████████████████████████| 1.4MB 14.4MB/s 
  Created wheel for spacy-wordnet: filename=spacy_wordnet-0.0.4-py2.py3-none-any.whl size=650292 sha256=13aadb7c684bb7162ef055850a9c30a1baa12962ff2fff62b50539e4dd0f192d
  Stored in directory: /root/.cache/pip/wheels/25/93/1d/c86db913cd146fc9ddb26d10f56579c5d58a3e00bc8f96a3a6
  Created wheel for nltk: filename=nltk-3.3-cp36-none-any.whl size=1394470 sha256=67f01727ce9fef3fa55a377ebcd489f985481953fd5257e8c7618fdcad51ab93
  Stored in directory: /root/.cache/pip/wheels/d1/ab/40/3bceea46922767e42986aef7606a600538ca80de6062dc266c
Successfully built spacy-wordnet nltk
  Found existing installation: nltk 3.2.5
    Uninstalling nltk-3.2.5:
      Successfully uninstalled nltk-3.2.5


## Import Libraries and initialize settings

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

In [0]:
import re
import string
import math
from tqdm import tqdm

In [0]:
import nltk
from nltk import pos_tag
from nltk.corpus import wordnet as wn
from nltk.tokenize import sent_tokenize, word_tokenize
from pprint import pprint
from nltk.tokenize.punkt import PunktSentenceTokenizer, PunktTrainer

In [0]:
import nltk
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('averaged_perceptron_tagger')
nltk.download('omw')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package omw to /root/nltk_data...
[nltk_data]   Package omw is already up-to-date!


True

In [0]:
import spacy
from spacy.matcher import Matcher 
from spacy.tokens import Span 
from spacy import displacy 
from spacy.lang.en.stop_words import STOP_WORDS
from spacy_wordnet.wordnet_annotator import WordnetAnnotator

nlp = spacy.load("en_core_web_lg")
nlp.add_pipe(WordnetAnnotator(nlp.lang), after='tagger')

In [0]:
import warnings
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
warnings.filterwarnings('ignore')

In [0]:
from sklearn.model_selection import train_test_split as tts 
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.metrics import confusion_matrix, recall_score, precision_score
from sklearn.preprocessing import StandardScaler

import lightgbm as lgb

from imblearn.over_sampling import SMOTE

In [0]:
from patsy import dmatrices
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

import pickle

## Define Functions

In [0]:
def get_target_data(path):
    df = pd.read_csv(path)
    df['cdc_present'] = (df['cdc'] != '---').astype(int)

    return df['cdc_present']

In [0]:
def topic_domains(topic):
    raw_domains = []
    common_domains = []

    for chunk in nlp(topic).noun_chunks:
        rt_nn = chunk.root
        raw_domains.append(rt_nn._.wordnet.wordnet_domains())

    if len(raw_domains) > 0:
        reference = raw_domains[0]
    else:
        return list(raw_domains)
        
    for domain in raw_domains:
        common_domains = set(reference) & set(domain)
        
    return list(common_domains)

In [0]:
def expanded_sentence_similarity(topic, sentence):
    
    extended_topic = []
    domains = topic_domains(topic)
    
    if len(domains) < 1:
        return nlp(topic).similarity(nlp(sentence))
    else:
        for token in nlp(topic):
            synsets = token._.wordnet.wordnet_synsets_for_domain(domains)
            if synsets: 
                lfs = []
                for syn in synsets:
                    lfs.extend(syn.lemma_names())
                    extended_topic.append('({})'.format('|'.join(set(lfs))))
            else:
                extended_topic.append(token.text)

        tokenized_extended_topic = nlp(' '.join(extended_topic))

    return tokenized_extended_topic.similarity(nlp(sentence))            

In [0]:
def subject_match(topic, sentence):
    
    ntopic = nlp(topic)
    scores = []
    for chunk in nlp(sentence).noun_chunks:
        scores.append(chunk.similarity(ntopic))
    
    return np.mean(scores)

In [0]:
def subject_match2(topic, sentence):
    
    ntopic = nlp(topic)
    scores = []
    for chnk in nlp(sentence).noun_chunks:
        for tchnk in ntopic.noun_chunks:
            scores.append(chnk.similarity(tchnk))
    
    return np.mean(scores)

In [0]:
def includes_that(sentence):
    check = []
    for tok in sentence:
        if tok.tag_ == 'IN':
            check.append(tok.text)
    if 'that' in check:
        return 1
    else:
        return 0

In [0]:
def patterns(sentence, pattern_type='dep'):
    
    pat = []
    if pattern_type == 'dep':
        pat = [tok.dep_ for tok in sentence]
    elif pattern_type == 'tag':
        pat = [tok.tag_ for tok in sentence]
    
    return '|'.join(pat)

In [0]:
def pos_pattern(sentence):
    pat = []
    for tok in sentence:
        pat.append(tok.pos_)

    return '|'.join(pat)


In [0]:
def pattern_in_text(pattern, sentence):
    text_pattern = pos_pattern(sentence)

    if pattern in text_pattern:
        return 1
    else:
        return 0

In [0]:
def infinitive_verb_present(sentence):
    text_pattern = pos_pattern(sentence)
    infinitive_pattern = 'PART|VERB'

    if infinitive_pattern in text_pattern:
        return 1
    else:
        return 0


In [0]:
def sentiment_subjectivity(sentence, score_type='sentiment'):
    analyzer = SentimentIntensityAnalyzer()

    if score_type == 'sentiment':
        pos = analyzer.polarity_scores(sentence)['pos']
        neg = analyzer.polarity_scores(sentence)['neg']
        return pos/neg
    
    else:
        return analyzer.polarity_scores(sentence)['compound']

In [0]:
def is_sentence(clause):
    nn = 2
    vb = 1
    cls = nlp(clause)
    for tok in cls:
        if tok.pos_ in ['NOUN', 'PROPN', 'PRON']:
            nn -= 1
        elif tok.pos_ in ['VERB']:
            vb -= 1
    if nn < 1 and vb < 1:
        return True
    else:
        return False

In [0]:
def get_clauses(sentence):
    """Sentence type is a string."""
    subtree_phrases = []
    
    for tok in nlp(sentence):
        if tok.head.pos_ == 'VERB':
            if len([tok for tok in tok.subtree]) >= 3:
                subtree_prhase = ''.join(tok.text_with_ws for tok in tok.subtree)
                if subtree_prhase != sentence:
                    if is_sentence(subtree_prhase):
                        splits = subtree_prhase.split(',')
                        if len(splits) == 1:
                            subtree_phrases.append(subtree_prhase)
                        else:
                            for split in splits:
                                if is_sentence(split):
                                    if split not in subtree_phrases:
                                        subtree_phrases.append(split) 
    
    raw_clauses = []
    clauses = []

    for phrase in subtree_phrases:
        tok_phrase = nlp(phrase)
        for chunk in tok_phrase.noun_chunks:
            span = tok_phrase[chunk.root.head.left_edge.i:chunk.root.head.right_edge.i+1]
            if span[0].pos_ in ['DET', 'ADP']:
                if span[1:].text not in raw_clauses:
                    if is_sentence(span[1:].text):
                        raw_clauses.append(span[1:].text)
            elif span.text not in raw_clauses:
                if is_sentence(span.text):
                    raw_clauses.append(span.text)
    
    if len(raw_clauses) == 0:
        clauses.append(sentence)
    else:
        for clause in raw_clauses:
            dup_checks = [(clause in check_clause) for check_clause in raw_clauses]
            if sum(dup_checks) == 1:
                clauses.append(clause)
                
    return clauses

In [0]:
def conf_matrix(y_true, y_predict):
    
    from sklearn.metrics import confusion_matrix
    
    data = confusion_matrix(y_true, y_predict)
    index = ['Actual_0', 'Actual_1']
    columns = ['Predicted_0', 'Predicted_1']
    
    return pd.DataFrame(data, index, columns)

In [0]:
def get_target_data(path):
    df = pd.read_csv(path)
    df['cdc_present'] = (df['cdc'] != '---').astype(int)

    return df['cdc_present']

In [0]:
def count_sentiment_words(sentence):
    analyzer = SentimentIntensityAnalyzer()
    wd_cnt = 0

    for tok in nlp(sentence):
        if analyzer.polarity_scores(tok.text)['compound'] != 0:
            wd_cnt += 1

    return wd_cnt

In [0]:
def count_syllables(sentence):
    """Given a string, calculates and returns the number of syllables of the word."""
    import string
    
    word_list = sentence.translate(string.punctuation).split()
    vowels = "aeiouy"
    tally = 0
    
    for word in word_list:
        word = word.lower()
        vowels = "aeiouy"
        count = 0
        
        if word[0] in vowels:
            count += 1

        for i in range(1, len(word)):
            if word[i] in vowels and word[i - 1] not in vowels:
                count += 1

        if word.endswith('e'):
            count -= 1

        if count == 0:
            count += 1
    
        tally += count
    
    return tally

In [0]:
def punct_marks(sentence):
    """Given a sentence, returns count and/or list of punctiation marks"""
    
    pm = []
    
    for tok in nlp(sentence):
        if tok.pos_ == 'PUNCT':
            pm.append(tok)
    return len(pm)

In [0]:
def noun_counts(sentence):
    
    chnk = []
    
    for chunk in nlp(sentence).noun_chunks:
        chnk.append(chunk)

    return len(chnk)

In [0]:
def word_count(sentence):
    wds = []
    for tok in nlp(sentence):
        if tok.pos_ != 'PUNCT':
            wds.append(tok)
    return len(wds)

In [0]:
def phrase_chunks(sentence):
    s = []
    for tok in nlp(sentence):
        if len([tok for tok in tok.subtree]) >=3:
            if ''.join(tok.text_with_ws for tok in tok.subtree) != str(sentence):
                s.append(''.join(tok.text_with_ws for tok in tok.subtree))
    return len(s)

## Load Dataframe

In [0]:
# paths to data
path = '/content/drive/My Drive/Colab Notebooks/Thinkful/Module 34 - Final Capstone/data/CDC Detection/cleaned.csv'

df = pd.read_csv(path)

In [0]:
df.head()

Unnamed: 0  Unnamed: 0.1  ... expanded_similarity subjectivity_score
0           0             0  ...            0.259504             0.0000
1           1             1  ...            0.267986            -0.4019
2           2             2  ...            0.106333             0.3182
3           3             3  ...            0.240641            -0.9280
4           4             4  ...            0.306673             0.3612

[5 rows x 21 columns]

In [0]:
# vectorize text
txt = [nlp(sentence) for sentence in df['text']]
v_txt = [t.vector for t in txt]

In [0]:
cxt = [nlp(sentence) for sentence in df['context']]
v_cxt = [t.vector for t in cxt]

## Preprocessing

In [0]:
txt_vects = pd.DataFrame(v_txt)
txt_vects.columns = ['txt_'+str(col) for col in txt_vects.columns]

cxt_vects = pd.DataFrame(v_cxt)
cxt_vects.columns = ['cxt_'+str(col) for col in cxt_vects.columns]

X = pd.concat([txt_vects, cxt_vects], axis=1)
y = df['cdc_present']

In [0]:
# Initialize SMOTE for class imbalance handling
sm = SMOTE(random_state=42)

# Split features,target into train/test sets
X_trn, X_tst, y_trn, y_tst = tts(txt_vects, y, test_size=0.25, random_state=42)

# resample train set
X_res, y_res = sm.fit_resample(X_trn, y_trn)


## Modeling

### Baseline

This initial model will use only vectorize full text and context features.

#### First pass

In [0]:
lr = LogisticRegression()
rf = RandomForestClassifier()
lb = lgb.LGBMClassifier()

lr.fit(X_res, y_res)
rf.fit(X_res, y_res)
lb.fit(X_res, y_res)

pn_lr = lr.predict(X_trn)
pt_lr = lr.predict(X_tst)
pn_rf = rf.predict(X_trn)
pt_rf = rf.predict(X_tst)
pn_lb = lb.predict(X_trn)
pt_lb = lb.predict(X_tst)

print("\n----------------------LR Scores----------------------")
print('Training set score:', lr.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_lr))
print('Training recall:', recall_score(y_trn, pn_lr))
print('\nTest set score:', lr.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_lr))
print('Test recall:', recall_score(y_tst, pt_lr))
print("\n----------------------LR Scores----------------------")
print('Training set score:', rf.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_rf))
print('Training recall:', recall_score(y_trn, pn_rf))
print('\nTest set score:', rf.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_rf))
print('Test recall:', recall_score(y_tst, pt_rf))
print("\n----------------------LR Scores----------------------")
print('Training set score:', lb.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_lb))
print('Training recall:', recall_score(y_trn, pn_lb))
print('\nTest set score:', lb.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_lb))
print('Test recall:', recall_score(y_tst, pt_lb))


----------------------LR Scores----------------------
Training set score: 0.8024064256008128
Training precision: 0.11493573644556064
Training recall: 0.7695351137487636

Test set score: 0.7951428571428572
Test precision: 0.1015625
Test recall: 0.7428571428571429

----------------------LR Scores----------------------
Training set score: 1.0
Training precision: 1.0
Training recall: 1.0

Test set score: 0.9614285714285714
Test precision: 0.3170731707317073
Test recall: 0.24761904761904763

----------------------LR Scores----------------------
Training set score: 0.9579669195847488
Training precision: 0.4281780633318036
Training recall: 0.9228486646884273

Test set score: 0.926
Test precision: 0.19117647058823528
Test recall: 0.45396825396825397


In [0]:
# Training set
conf_matrix(y_trn, pn_lr)

Predicted_0  Predicted_1
Actual_0        24497         5991
Actual_1          233          778

In [0]:
# Test set
conf_matrix(y_tst, pt_lr)

Predicted_0  Predicted_1
Actual_0         8115         2070
Actual_1           81          234

In [0]:
# Training set
conf_matrix(y_trn, pn_rf)

Predicted_0  Predicted_1
Actual_0        30488            0
Actual_1            0         1011

In [0]:
# Test set
conf_matrix(y_tst, pt_rf)

Predicted_0  Predicted_1
Actual_0        10017          168
Actual_1          237           78

In [0]:
# Training set
conf_matrix(y_trn, pn_lb)

Predicted_0  Predicted_1
Actual_0        29242         1246
Actual_1           78          933

In [0]:
# Test set
conf_matrix(y_tst, pt_lb)

Predicted_0  Predicted_1
Actual_0         9580          605
Actual_1          172          143

### Adding Features

At this point, recall of positive instances is more important than overall model accuracy.

Logistic Regression, with a recall score of .743, easily outperformed other models.  Additionally, it generalizes well when compared to its training scores.  Given how distinct the advantage Logistic Regression has over other models, in both overfitting and recall, it will likely be the selected model for this task.  

The next model will evaluate the use of vectorized full text, vectorized context, and the addition of other engineered features.

In [0]:
sc = StandardScaler()
F = pd.DataFrame()

F['txt_wds'] = df['text_words']
F['txt_syl'] = df['text_syllables']
F['snt_rat'] = df['text_sentiment_words'] / df['text_words']
F['txt_pct'] = df['text_punctuation']
F['txt_nns'] = df['text_nouns']
F['txt_phs'] = df['text_phrases']
F['tpc_sim'] = df['context_similarity']
F['exp_sim'] = df['expanded_similarity']
F['subj_sc'] = df['subjectivity_score']

In [0]:
sc_F = sc.fit_transform(F)

sc_F = pd.DataFrame(sc_F)

In [0]:
X = pd.concat([X, sc_F], axis=1)

In [0]:
X_trn, X_tst = X.iloc[X_trn.index], X.iloc[X_tst.index]

X_res, y_res = sm.fit_resample(X_trn, y_trn)

In [0]:
lr = LogisticRegression()
rf = RandomForestClassifier()
lb = lgb.LGBMClassifier()

lr.fit(X_res, y_res)
rf.fit(X_res, y_res)
lb.fit(X_res, y_res)

pn_lr = lr.predict(X_trn)
pt_lr = lr.predict(X_tst)
pn_rf = rf.predict(X_trn)
pt_rf = rf.predict(X_tst)
pn_lb = lb.predict(X_trn)
pt_lb = lb.predict(X_tst)

print("\n----------------------LR Scores----------------------")
print('Training set score:', lr.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_lr))
print('Training recall:', recall_score(y_trn, pn_lr))
print('\nTest set score:', lr.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_lr))
print('Test recall:', recall_score(y_tst, pt_lr))
print("\n----------------------RF Scores----------------------")
print('Training set score:', rf.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_rf))
print('Training recall:', recall_score(y_trn, pn_rf))
print('\nTest set score:', rf.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_rf))
print('Test recall:', recall_score(y_tst, pt_rf))
print("\n----------------------LB Scores----------------------")
print('Training set score:', lb.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_lb))
print('Training recall:', recall_score(y_trn, pn_lb))
print('\nTest set score:', lb.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_lb))
print('Test recall:', recall_score(y_tst, pt_lb))


----------------------LR Scores----------------------
Training set score: 0.8056128765992571
Training precision: 0.1221170904790065
Training recall: 0.8170128585558852

Test set score: 0.8044761904761905
Test precision: 0.1099640933572711
Test recall: 0.7777777777777778

----------------------RF Scores----------------------
Training set score: 1.0
Training precision: 1.0
Training recall: 1.0

Test set score: 0.9664761904761905
Test precision: 0.39664804469273746
Test recall: 0.2253968253968254

----------------------LB Scores----------------------
Training set score: 0.9827931045430014
Training precision: 0.7994891443167306
Training recall: 0.619188921859545

Test set score: 0.9661904761904762
Test precision: 0.38372093023255816
Test recall: 0.20952380952380953


In [0]:
# Training set
conf_matrix(y_trn, pn_lr)

Predicted_0  Predicted_1
Actual_0        24550         5938
Actual_1          185          826

In [0]:
# Test set
conf_matrix(y_tst, pt_lr)

Predicted_0  Predicted_1
Actual_0         8202         1983
Actual_1           70          245

In [0]:
# Training set
conf_matrix(y_trn, pn_rf)

Predicted_0  Predicted_1
Actual_0        30488            0
Actual_1            0         1011

In [0]:
# Test set
conf_matrix(y_tst, pt_rf)

Predicted_0  Predicted_1
Actual_0        10077          108
Actual_1          244           71

In [0]:
# Training set
conf_matrix(y_trn, pn_lb)

Predicted_0  Predicted_1
Actual_0        30331          157
Actual_1          385          626

In [0]:
# Test set
conf_matrix(y_tst, pt_lb)

Predicted_0  Predicted_1
Actual_0        10079          106
Actual_1          249           66

### Tuning the LR model

In [0]:
params = [{'C':[.001, .01, .1, 1, 10, 100, 1000, 1000],
           'penalty': ['l1', 'l2']}]

lr = LogisticRegression()

clf = GridSearchCV(lr, param_grid=params, scoring='recall', cv=3)
clf.fit(X_res, y_res)


GridSearchCV(cv=3, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=None, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid=[{'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000, 1000],
                          'penalty': ['l1', 'l2']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='recall', verbose=0)

In [0]:
clf.best_estimator_

LogisticRegression(C=1, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [0]:
lr = LogisticRegression(C=1, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)
lr.fit(X_res, y_res)

pn_lr = lr.predict(X_trn)
pt_lr = lr.predict(X_tst)

print("\n----------------------LR Scores----------------------")
print('Training set score:', lr.score(X_trn, y_trn))
print('Training precision:', precision_score(y_trn, pn_lr))
print('Training recall:', recall_score(y_trn, pn_lr))
print('\nTest set score:', lr.score(X_tst, y_tst))
print('Test precision:', precision_score(y_tst, pt_lr))
print('Test recall:', recall_score(y_tst, pt_lr))


----------------------LR Scores----------------------
Training set score: 0.8056128765992571
Training precision: 0.1221170904790065
Training recall: 0.8170128585558852

Test set score: 0.8044761904761905
Test precision: 0.1099640933572711
Test recall: 0.7777777777777778


In [0]:
# Training set
conf_matrix(y_trn, pn_lr)

Predicted_0  Predicted_1
Actual_0        24550         5938
Actual_1          185          826

In [0]:
# Test set
conf_matrix(y_tst, pt_lr)

Predicted_0  Predicted_1
Actual_0         8202         1983
Actual_1           70          245

In [0]:
scores = cross_val_score(lr, X_res, y_res, cv=5)

In [0]:
print(scores)

[0.82732043 0.8400984  0.82542025 0.83714637 0.83583436]


This model will work serve as the first funnel.

In [0]:
X.head()

txt_0     txt_1     txt_2  ...         6         7         8
0 -0.043492  0.255405 -0.031507  ... -0.108448 -0.258628 -0.019056
1 -0.109559  0.207248  0.102608  ...  0.203428 -0.188807 -1.041112
2 -0.112302  0.123262  0.083034  ... -0.503771 -1.519513  0.790146
3 -0.053385  0.214746  0.021882  ...  0.786688 -0.413902 -2.379016
4 -0.018335  0.202566  0.023854  ... -0.163275  0.129662  0.899497

[5 rows x 609 columns]

In [0]:
file = '/content/drive/My Drive/Colab Notebooks/Thinkful/Module 34 - Final Capstone/data/CDC Detection/funnel_1_model.sav'

pickle.dump(lr, open(file, 'wb'))

In [0]:
save_df = pd.concat([X_tst,y_tst], axis=1)

write_path = '/content/drive/My Drive/Colab Notebooks/Thinkful/Module 34 - Final Capstone/data/CDC Detection/m1_test_set.csv'

save_df.to_csv(write_path)